### PR TITLE
Type inference in conditional types

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR introduces the ability to use type inference in conditional types (#21316), enabling *pattern matching for types*. For example, the following extracts the return type of a function type:

```ts
type ReturnType<T> = T extends (...args: any[]) => infer R ? R : T;
```

Within the `extends` clause of a conditional type, it is now possible to have `infer` declarations that introduce a type variable to be inferred. Such inferred type variables may be referenced in the true branch of the conditional type. It is possible to have multiple `infer` locations for the same type variable.

A conditional type `T extends U ? X : Y` is either *resolved* to `X` or `Y`, or *deferred* because the condition depends on one or more type variables. Whether to resolve or defer is determined as follows:

* First, given types `T'` and `U'` that are instantiations of `T` and `U` where all occurrences of type parameters are replaced with `any`, if `T'` is not assignable to `U'`, the conditional type is resolved to `Y`. Intuitively, if the most permissive instantiation of `T` is not assignable to the most permissive instantiation of `U`, we know that no instantiation will be and we can just resolve to `Y`.
* Next, for each type variable introduced by an `infer` declaration within `U` collect a set of candidate types by inferring from `T` to `U` (using the same inference algorithm as type inference for generic functions). For a given `infer` type variable `V`, if any candidates were inferred from co-variant positions, the type inferred for `V` is a union of those candidates. Otherwise, if any candidates were inferred from contra-variant positions, the type inferred for `V` is an intersection of those candidates. Otherwise, the type inferred for `V` is `never`.
* Then, given a type `T''` that is an instantiation of `T` where all `infer` type variables are replaced with the types inferred in the previous step, if `T''` is *definitely assignable* to `U`, the conditional type is resolved to `X`. The definitely assignable relation is the same as the regular assignable relation, except that type variable constraints are not considered. Intuitively, when a type is definitely assignable to another type, we know that it will be assignable for *all instantiations* of those types.
* Otherwise, the condition depends on one or more type variables and the conditional type is deferred.

Conditional types can be nested to form a sequence of pattern matches that are evaluated in order:

```ts
type Unpacked<T> =
    T extends (infer U)[] ? U :
    T extends (...args: any[]) => infer U ? U :
    T extends Promise<infer U> ? U :
    T;

type T0 = Unpacked<string>;  // string
type T1 = Unpacked<string[]>;  // string
type T2 = Unpacked<() => string>;  // string
type T3 = Unpacked<Promise<string>>;  // string
type T4 = Unpacked<Promise<string>[]>;  // string[]
type T5 = Unpacked<Unpacked<Promise<string>[]>>;  // string
```

Note that is not possible for a conditional type to recursively reference itself, as might be desired in the `Unpacked<T>` case above. We're still considering ways in which to implement this.

The following example demonstrates how multiple candidates for the same type variable in co-variant positions causes a union type to be inferred:

```ts
type Foo<T> = T extends { a: infer U, b: infer U } ? U : never;
type T10 = Foo<{ a: string, b: string }>;  // string
type T11 = Foo<{ a: string, b: number }>;  // string | number
```

Likewise, multiple candidates for the same type variable in contra-variant positions causes an intersection type to be inferred:

```ts
type Bar<T> = T extends { a: (x: infer U) => void, b: (x: infer U) => void } ? U : never;
type T20 = Bar<{ a: (x: string) => void, b: (x: string) => void }>;  // string
type T21 = Bar<{ a: (x: string) => void, b: (x: number) => void }>;  // string & number
```

When inferring from a type with multiple call signatures (such as the type of an overloaded function), inferences are made from the *last* signature (which, presumably, is the most permissive catch-all case). It is not possible to perform overload resolution based on a list of argument types (this would require us to support `typeof` for arbitrary expressions, as suggested in #6606, or something similar).

```ts
declare function foo(x: string): number;
declare function foo(x: number): string;
declare function foo(x: string | number): string | number;
type T30 = ReturnType<typeof foo>;  // string | number
```

It is not possible to use `infer` declarations in constraint clauses for regular type parameters:

```ts
type ReturnType<T extends (...args: any[]) => infer R> = R;  // Error, not supported
```

However, much the same effect can be obtained by erasing the type variables in the constraint and instead specifying a conditional type:

```ts
type AnyFunction = (...args: any[]) => any;
type ReturnType<T extends AnyFunction> = T extends (...args: any[]) => infer R ? R : never;
```

### CODE REVIEW

**Comment by phiresky** [comment]
oh lord this is amazing please merge

**Comment by nevir** [comment]
Can't wait to write a JS to TS-Type compiler; why run your code when TS can just infer it! :P

**Comment by jack-williams** [comment]
Question regarding second bullet point. If an inference variable appears in _both_ contravariant and covariant positions does the catch-all case apply and _always_ infer `never`?

**Comment by ahejlsberg** [comment]
@jack-williams Covariant candidates take precedence over contravariant candidates, so when both are present we make an inference only from the covariant candidates. We resort to `never` only when there are no candidates at all.

**Comment by kastermester** [comment]
Given the following code:

```
declare class MyClass {}

type InstanceType<T> = T extends new (...args: any[]): infer U ? U : never;

type MyClassInstance = InstanceType<typeof MyClass>
```

Would that work? Ie get the instance type from a constructor type / inverse of typeof (when applied to constructor functions)

**Comment by rictic** [comment]
@kastermester Just checked, gotta wrap the `new ()` part in `{}` but otherwise that works as written. `MyClassInstance` is `MyClass`.

**Comment by jack-williams** [comment]
Thanks for the reply @ahejlsberg! I think you can also use this to query a function's inputs, cool!

```typescript
interface FnShape<N> {
    0: (x: N, ...args: any[]) => any;
    1: (x: any, y: N, ...args: any[]) => any;
    2: (x: any, y: any, z: N, ...args: any[]) => any;
    Return: (...args: any[]) => N;
}

type QueryFn<N extends keyof FnShape<any>, T extends FnShape<any>[N]> = T extends FnShape<infer R>[N]?R: T

type Foo = (x: number, y: boolean, z: string) => void

type X = QueryFn<"0", Foo>;
type Y = QueryFn<"1", Foo>;
type Z = QueryFn<"2", Foo>;
type Ret = QueryFn<"Return", Foo>;
```

**Comment by rictic** [comment]
I'm seeing a `tsc` crash with taking the intersection type of the return type of a mixin function. `InstanceType` is more elegant anyways, and it works, but I figured I should report the crash here regardless:

```typescript
type Constructor<T = {}> = {new (...args: any[]): T};

function Timestamped<TBase extends Constructor>(Base: TBase) {
    return class extends Base {
        timestamp = Date.now();
    };
}
type Timestamped = ReturnTypeOf<typeof Timestamped>

class Foo {
  foo = 'foo';
}

const TimestampedFoo = Timestamped(Foo);
type InstanceType<T> = T extends new (...args: any[]) => infer U ? U : never;
// This works:
type TimestampedFoo = InstanceType<typeof TimestampedFoo>;

// This crashes tsc with the message:
//     Cannot read property 'get' of undefined
// type TimestampedFoo = Foo & Timestamped;


const timestampedFoo: TimestampedFoo = new TimestampedFoo();
const string: string = timestampedFoo.foo;
const date: number = timestampedFoo.timestamp;
```

File: src/compiler/types.ts [hunk 1]
```diff
@@ -3833,6 +3883,8 @@ namespace ts {
         NakedTypeVariable = 1 << 0,  // Naked type variable in union or intersection type
         MappedType        = 1 << 1,  // Reverse inference for mapped type
         ReturnType        = 1 << 2,  // Inference made from return type of generic function
+        NoConstraints     = 1 << 3,  // Don't infer from constraints of instantiable types
```
Comment (mhegazy) [inline src/compiler/types.ts]: kinda strange to have these on priority, and not on inferenceInfo for instance.

File: src/compiler/types.ts [hunk 2]
```diff
@@ -3833,6 +3883,8 @@ namespace ts {
         NakedTypeVariable = 1 << 0,  // Naked type variable in union or intersection type
         MappedType        = 1 << 1,  // Reverse inference for mapped type
         ReturnType        = 1 << 2,  // Inference made from return type of generic function
+        NoConstraints     = 1 << 3,  // Don't infer from constraints of instantiable types
+        AlwaysStrict      = 1 << 4,  // Always use strict rules for contravariant inferences
```
Comment (mhegazy) [inline src/compiler/types.ts]: i am not sure i understand that one really.. i think it might be easier to check with you in person.

**Comment by mhegazy** [comment]
@andy-ms there are a couple of things we need to validate for this change, e.g. adding formatting rules for the new type syntax, and making sure the infered type parameter shows up in completion in the true branch.

**Comment by Igorbek** [comment]
@ahejlsberg in the examples above, did you actually mean
```ts
type T4 = Unpacked<Promise<string>[]>;  // Promise<string>, not string[]?
```

Also, is there a reason why it is limited to only _conditional types_? I'm sure that there are cases where we want to enforce/constraint that type is actually in the true branch and condition matches always.

For example,
```ts
// why do we care of false branch?
type ReturnType1<F extends (...args: any[]) => any> = F extends (...args: any[]) => infer R ? R : T;

// isn't it going to better?
type ReturnType1<F extends (...args: any[]) => infer R> = R;
```

**Comment by rbuckton** [comment]
@ahejlsberg I noticed that `type T = never extends never ? true : false;` now works, but it looks like this doesn't:

infertypes

**Comment by ahejlsberg** [comment]
@Igorbek Yes, thanks for the correction. Now fixed.

We may at a later date support `infer` declarations in regular type parameter constraints, however it is not trivial to implement. Meanwhile, you can get much the same effect with the existing constraint mechanism as I explain in the introduction to the PR.

**Comment by ahejlsberg** [comment]
@rictic Good catch, thanks for pointing this out. Fix is up in #21610.

**Comment by jcalz** [comment]
We've been able to programmatically determine the union of values of a type for a while now:
```ts
// union of all value types
type ValueOf<T> = T[keyof T]; 
```

But this looks like the first time we can do the same for the *intersection* of values, by taking advantage of the contravariance of function parameters:
```ts
// intersection of all value types
type IntersectValueOf<T> =
  ({[K in keyof T]: (x: T[K]) => void}) extends Record<keyof T, (x: infer V) => void> ? V: never
```

Demo:
```ts
type Thing = {
  foo: {a: string},
  bar: {b: number},
  baz: {c: boolean}
}

type Something = ValueOf<Thing>;
declare const something: Something;
if ('a' in something) something.a; // string
if ('b' in something) something.b; // number
if ('c' in something) something.c; // boolean

type Everything = IntersectValueOf<Thing>;
declare const everything: Everything;
everything.a; // string
everything.b; // number
everything.c; // boolean
```

Once `infer` makes it to a release, can we rely on something like the above working?  Or is that too crazy?

**Comment by phiresky** [comment]
I think I found a bug:

I'm trying to create a type that would correctly represent this function:

```ts
function sanitize(x: any): any {
    return JSON.stringify(JSON.parse(x));
}
```
I.e., primitives should stay, `Date` should become `string` and functions should be removed. (Very useful for stuff sent over the wire)

It should look pretty much like this:

```ts
type Jsonified<T> = 
    T extends string | number | boolean | null ? T 
    : T extends undefined | Function ? never // undefined and functions are removed
    : T extends {toJSON(): infer R} ? R // toJSON is called if it exists (e.g. Date)
    : T extends object ? {[k in keyof T]: JSONified<T[k]>}
    : "what is this";
```

This almost works, except for the recursion:

```ts
type example = {
    str: "literalstring",
    fn: () => void,
    date: Date,
    customClass: MyClass,
    obj: {
        prop: "property",
        nested: { attr: Date }
    },
}
declare class MyClass {
    toJSON(): "correct";
}
```

should give

```ts
type jsonifiedExample = Jsonified<example>;
type jsonifiedExample == {
    str: "literalstring";
    fn: never;
    date: string;
    customClass: "correct";
    obj: { prop: "property"; nested: { attr: string; } }
}
```

but it actually returns
```ts
type jsonifiedExample == {
    str: "literalstring";
    fn: never;
    date: string;
    customClass: "correct";
    obj: any;
}
```

I can make it go one level deeper by wrapping the `extends object` part in Jsonified twice, but not for any level of nesting.

Additionally, it doesn't seem possible to recurse on inferred variables (`T extends {toJSON(): infer R} ? Jsonifiied<R>` does not work)

**Comment by gcanti** [comment]
@phiresky The `obj: any;` part looks related to this bug https://github.com/Microsoft/TypeScript/issues/14041

As a workaround you can define a type alias 

```ts
type JsonifiedObject<T> = {[k in keyof T]: Jsonified<T[k]>}
type Jsonified<T> =
    T extends string | number | boolean | null ? T
    : T extends undefined | Function ? never // undefined and functions are removed
    : T extends {toJSON(): infer R} ? R // toJSON is called if it exists (e.g. Date)
    : T extends object ? JsonifiedObject<T> // <= type alias here
    : "what is this";
```

**Comment by phiresky** [comment]
@gcanti 

Right, my issue seems to also only apply to what is shown when hovering, when initializing a variable of that type the property suggestions are correct. Still annoying, but far better :)

Your example is even more broken though, even though it looks correct: It gives this type

```ts
type jsonifiedExample = {
    str: "literalstring";
    num: 123;
    nll: null;
    fn: () => void;
    date: Date;
    customClass: MyClass;
    obj: {
        prop: "property";
        nested: {
            attr: Date;
        };
    };
}
```

-> the type mapping is not applied at all

**Comment by jcalz** [comment]
@ahejlsberg [wrote](https://github.com/Microsoft/TypeScript/pull/21496#issue-292940972):
> Note that is not possible for a conditional type to recursively reference itself, as might be desired in the `Unpacked<T>` case above. We're still considering ways in which to implement this.

**Comment by ahejlsberg** [comment]
@phiresky The issue you're seeing above is indeed a bug. Easy fix, I will have a PR up soon.

**Comment by ahejlsberg** [comment]
@phiresky Fix is up in #21696.

**Comment by ahejlsberg** [comment]
@phiresky BTW, the reason you're seeing `obj: any` for the nested object is that our quick info type-to-string conversion only goes one level deep in recursive types. If you actually access the property in an expression you'll see that it has the correct type.

**Comment by simonbuchan** [comment]
Does it make sense to add a constraint to the inferred type? I'm trying to thing of a useful case that can't already be done.

Hmm, purely syntactically, `infer` looks like it should extend to inline generic parameter introduction:

```ts
function zip(a: ReadonlyArray<infer A>, b: ReadonlyArray<infer B>) => Array<[A, B]>;
```

How much does that make the team throw up in their mouths 😉?

**Comment by MadaraUchiha** [comment]
@simonbuchan You don't need `infer` for that,

```typescript
declare function zip<A, B>(a: ReadonlyArray<A>, b: ReadonlyArray<B>): Array<[A, B]>;
```

You really only need infer when A is part of another type you don't particularly know about, for example, you have `T` which could be a `U[]`, and you need `U`.

**Comment by simonbuchan** [comment]
The issue is more about how you would describe `infer` so that users don't expect to be able to do that. (Or maybe decide that alternate syntax is fine and implement it, I dunno.)

Basically: `infer T` means something like "introduce a new type parameter that is inferred to the contextual type apon instantiation", but that definition is too wide.

Here's another case, where the user is trying to emulate the flow `*` type:

```ts
// Must be a promise, but I can't be bothered writing what kind
let foo: Promise<infer T> = asyncMethod();
```

**Comment by benstevens48** [comment]
I actually suggested this more general use of `infer` (with the syntax `=T` instead of `infer T`) some time ago in https://github.com/Microsoft/TypeScript/issues/13949. It will be interesting to see if this is implemented at some point. The current implementation is already very useful though, thanks!

**Comment by masaeedu** [comment]
it's happening!

**Comment by tycho01** [comment]
@jcalz: I can't tell what JS function you're typing there, but now that `infer` is in, I tried to run your type. It doesn't seem to be working yet though. I'm getting this:
```ts
type Everything = {
    foo: (x: {
        a: string;
    }) => void;
    bar: (x: {
        b: number;
    }) => void;
    baz: (x: {
        c: boolean;
    }) => void;
}
```
.. so apparently these `Everything` instances don't directly have `a` / `b` / `c`.

**Comment by jcalz** [comment]
@tycho01 

Hmm, I just checked 2.8.0-dev.20180211 and it gives:

```ts
type Everything = {
    a: string;
} & {
    b: number;
} & {
    c: boolean;
}
```
which is indeed the intersection type I was aiming for.   Not sure why you see something else.

**Comment by MeirionHughes** [comment]
a welcome addition; but don't you think `ReturnType<T>` is confusing? I would expect it to return the actual typescript Type of some function T. Perhaps `ReturnPrimitive<T>` would allow for a future `ReturnType` that can be used for currying?

**Comment by masaeedu** [comment]
@MeirionHughes `ReturnType` does "return" the return type of a TypeScript function type. Perhaps you're talking about a runtime representation of a type, which at least currently against the stated goals of the project (see #1573).

**Comment by MeirionHughes** [comment]
Yeah, my apologies - I was confusing this one with 21316.

**Comment by ericanderson** [comment]
Any updates on recursively handling this?

**Comment by tycho01** [comment]
@ericanderson https://github.com/Microsoft/TypeScript/pull/21613#issuecomment-365774702

**Comment by MeirionHughes** [comment]
I'm trying to do a promisify and it seems to die if you give it an overloaded method. 

The single case works perfectly fine: 
```
type Promisify<T> =
  T extends (cb: (v: infer V) => void) => void ? () => Promise<V> :
  T extends (a: infer A, cb: (v: infer V) => void) => void ? (a: A) => Promise<V> :
  T extends (a: infer A, b: infer B, cb: (v: infer V) => void) => void ? (a: A, b: B) => Promise<V> :
  T;

type Foo = ((a: string, cb: (v) => void) => void)

let bar: Promisify<Foo>;
```

overload it and it dies completely: 

```ts
type Foo =
  ((a: string, cb: (v) => void) => void) &
  ((a: string, b: number, cb: (v) => void) => void);
```

bug? already reported?

**Comment by MeirionHughes** [comment]
@ahejlsberg this limitation on the function overloads is super annoying - I can't seem to find a way to "mask" the input function (pick from the overloads) either. 

You mention you choose the last function from the available ones. Would it not be possible to try each until you arrive at a non-`never` result?

**Pull request merged**